File: cleanCode/二叉树/遍历/层序遍历/107. 二叉树的层次遍历 II.go
package 层序遍历

func levelOrderBottom(root *TreeNode) [][]int {
	return reverse(getLevelOrder(root))
}

func getLevelOrder(root *TreeNode) [][]int {
	sequences := make([][]int, 0)
	queue := NewTreeNodeQueue()
	queue.Push(root)
	for queue.IsNotEmpty() {
		sequenceOfTCurrentLay := make([]int, 0)
		for countOfUnRecordNodes := queue.GetSize(); countOfUnRecordNodes > 0; countOfUnRecordNodes-- {
			top := queue.Pop()
			if top == nil {
				continue
			}
			sequenceOfTCurrentLay = append(sequenceOfTCurrentLay, top.Val)
			queue.Push(top.Left, top.Right)
		}
		if len(sequenceOfTCurrentLay) != 0 {
			sequences = append(sequences, sequenceOfTCurrentLay)
		}
	}
	return sequences
}

func reverse(sequences [][]int) [][]int {
	length := len(sequences)
	for i := 0; i < length/2; i++ {
		sequences[i], sequences[length-1-i] = sequences[length-1-i], sequences[i]
	}
	return sequences
}

// --------------------------- TreeNodeQueue ---------------------------
type TreeNodeQueue struct {
	data []*TreeNode
}

func NewTreeNodeQueue() *TreeNodeQueue {
	return &TreeNodeQueue{}
}

func (tnq *TreeNodeQueue) Push(treeNodes ...*TreeNode) {
	tnq.data = append(tnq.data, treeNodes...)
}

func (tnq *TreeNodeQueue) Pop() *TreeNode {
	top := tnq.data[0]
	tnq.data = tnq.data[1:]
	return top
}

func (tnq *TreeNodeQueue) IsNotEmpty() bool {
	return tnq.GetSize() != 0
}

func (tnq *TreeNodeQueue) GetSize() int {
	return len(tnq.data)
}

/*
	题目链接;
	总结:
		1. 这题和 _102. 二叉树的层序遍历_ 差不多，求出层序遍历序列后翻转就可以了。
*/

File: cpp/queue/107. 二叉树的层次遍历 II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        queue<TreeNode*> queue;
        vector<vector<int>> layerVectors;
        queue.push(root);
        while(!queue.empty()){ 
            int size = queue.size();
            vector<int> layerVector;
            while(size--){
                TreeNode* top = queue.front();
                queue.pop();
                if (top == NULL){
                    continue;
                }
                layerVector.push_back(top->val);
                queue.push(top->left);
                queue.push(top->right);
            }
            if(layerVector.size()!=0){
                layerVectors.push_back(layerVector);
            }
        }
        reverse(layerVectors.begin(),layerVectors.end());
        return layerVectors;
    }
};

/*
    知识点
    1. 非空指针的判定。
    2. queue 的使用。
    3. vector 的使用。
*/